lib/studies/reply/app.dart

```diff
@@ -44,38 +66,86 @@ class ReplyApp extends StatelessWidget {
-        debugShowCheckedModeBanner: false,
-        theme: replyTheme,
-        localizationsDelegates: GalleryLocalizations.localizationsDelegates,
-        supportedLocales: GalleryLocalizations.supportedLocales,
-        locale: GalleryOptions.of(context).locale,
-        initialRoute: homeRoute,
-        onGenerateRoute: (settings) {
-          switch (settings.name) {
-            case homeRoute:
-              return MaterialPageRoute<void>(
-                builder: (context) => const AdaptiveNav(),
-                settings: settings,
-              );
-              break;
-            case composeRoute:
-              return createComposeRoute(settings);
-              break;
-          }
-          return null;
-        },
+    return RestorationScope(
+      restorationId: 'replyAppState',
+      child: MultiProvider(
+        providers: [
+          ChangeNotifierProvider<EmailStore>.value(
+            value: _appState.value,
+          ),
+        ],
+        child: MaterialApp(
+          navigatorKey: rootNavKey,
+          restorationScopeId: 'appNavigator',
+          title: 'Reply',
+          debugShowCheckedModeBanner: false,
+          theme: replyTheme,
+          localizationsDelegates: GalleryLocalizations.localizationsDelegates,
+          supportedLocales: GalleryLocalizations.supportedLocales,
+          locale: GalleryOptions.of(context).locale,
+          initialRoute: ReplyApp.homeRoute,
+          onGenerateRoute: (settings) {
+            switch (settings.name) {
+              case ReplyApp.homeRoute:
+                return MaterialPageRoute<void>(
+                  builder: (context) => const AdaptiveNav(),
+                  settings: settings,
+                );
+                break;
+              case ReplyApp.composeRoute:
+                return ReplyApp.createComposeRoute(settings);
+                break;
+            }
+            return null;
+          },
+        ),
       ),
     );
   }
 }
 
+class _RestorableEmailState extends RestorableListenable<EmailStore> {
+  @override
+  EmailStore createDefaultValue() {
+    return EmailStore();
+  }
+
+  @override
+  EmailStore fromPrimitives(Object data) {
+    final appState = EmailStore();
+    final appData = Map<String, dynamic>.from(data as Map);
+    appState.selectedEmailId = appData['selectedEmailId'] as int;
+    appState.onSearchPage = appData['onSearchPage'] as bool;
+
+    // The index of the MailboxPageType enum is restored.
+    final mailboxPageIndex = appData['selectedMailboxPage'] as int;
+    appState.selectedMailboxPage = MailboxPageType.values[mailboxPageIndex];
+
+    final starredEmailIdsList = appData['starredEmailIds'] as List<dynamic>;
+    appState.starredEmailIds = {
+      ...starredEmailIdsList.map<int>((dynamic id) => id as int),
+    };
+    final trashEmailIdsList = appData['trashEmailIds'] as List<dynamic>;
+    appState.trashEmailIds = {
+      ...trashEmailIdsList.map<int>((dynamic id) => id as int),
+    };
```

Comment:
nit: I think this can be just:

```dart
    appState.starredEmailIds = (appData['starredEmailIds'] as List<int>).toSet();
    appState.trashEmailIds = (appData['trashEmailIds'] as List<int>).toSet();
```

Reply:
When I tried this, I get a _CastError:
```
══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following _CastError was thrown building MediaQuery(MediaQueryData(size: Size(800.0, 600.0),
devicePixelRatio: 3.0, textScaleFactor: 1.0, platformBrightness: Brightness.light, padding:
EdgeInsets.zero, viewPadding: EdgeInsets.zero, viewInsets: EdgeInsets.zero, alwaysUse24HourFormat:
false, accessibleNavigation: false, highContrast: false, disableAnimations: false, invertColors:
false, boldText: false, navigationMode: traditional)):
type 'List<dynamic>' is not a subtype of type 'List<int>' in type cast

The relevant error-causing widget was:
  MediaQuery file:///Users/shihaohong/dev/apps/gallery/lib/data/gallery_options.dart:180:21

When the exception was thrown, this was the stack:
#0      _RestorableEmailState.fromPrimitives (package:gallery/studies/reply/app.dart:124:60)
#1      RestorationMixin.registerForRestoration (package:flutter/src/widgets/restoration.dart:845:20)
#2      _ReplyAppState.restoreState (package:gallery/studies/reply/app.dart:50:5)
```

I find type casting from dynamic to be pretty confusing and haven't found a sane way to determine if I can cast something a certain way other than trial and error. Any more ideas for this one?

Reply:
Oh strange, I don't have an idea. Let's drop this then, it does not seem worth to simplify it.